(FULL) Easybuild update integration

NeutryFD opened the pull request:
Add  new options  --complete-dependencies  for R packages 
- Update version and checksum from  CRAN database to easyconfig file 
- Write  dependecies tree of extension into easyconfig file with the correct order for the installation
-  Create a .bk file of the original easyconfig file

NeutryFD committed:
add --complete-dependencies option to the easybuild-framework, missing outpur format

NeutryFD committed:
rewrite config file with the new imports by order of installation

NeutryFD committed:
add  comments

Commit:
Added .venv and removed typo in .gitignore

Commit:
Complete_dependencies refactored. Several TODO's pending to be answered

Commit:
Add TODO regarding original extensions version

macheitor commented:
Improved NeutryFD's version of complete_dependencies. 
There are several TODO's pending to be solved by someone with more experience in EasyBuild

Commit:
Added fix when dependency has no imports + comments

Commit:
Refactor. Better abstraction of modules. Few TODO pending to implement for 'R' language

Commit:
Complete dependencies for R completed

Commit:
Delete unneded sorting function and avoid processing the same package twice

Commit:
Complete exts_list refactored with classes

Commit:
Added basic structure for PythonExtension download packages

Commit:
Support for python pypi packages. missing conda

Commit:
Added comment on dev instructions

Commit:
Update comments

Commit:
Now we check for the version that complies with all the version constraints . PythonExtensions needed to be adapted too

Commit:
complete_exts_list for R completed

Commit:
Finished also for python

macheitor commented:
Changed flag to "--complete-exts-list" which is literally what we are doing. 
We are not updating the version of the packages already present on the exts_list. 
This would be another feature, also is really easy to do this on the current code.

It is working now for R, R-Bioconductor and Python

Commit:
--update-exts-list feature added

dagonzalezfo commented on easybuild/framework/easyblock.py:
```diff
@@ -4835,3 +4845,1025 @@ def make_checksum_lines(checksums, indent_level):
         write_file(ec['spec'], ectxt)
+
+
+def exts_list_utils(ecs, complete=False, update=False):
```
This exts_utils can be moved to tools.py

dagonzalezfo commented on easybuild/framework/easyblock.py:
```diff
@@ -4835,3 +4845,1025 @@ def make_checksum_lines(checksums, indent_level):
+                print_error(f"Failed to get complete exts list: {e}", log=_log)
+                return None
+        
+    class PythonExtension:
```
Check available EB clases to avoid clashes/redundancies

dagonzalezfo reviewed (commented):
Do changes as proposed in meet session:
Use available EB clases if available and add funcions if needed overcharge it by adding new funcionalities.

Move ext related tooling (updates and other magic) into tools.py.

dagonzalezfo commented:
By exploring with Victor where is the best way to add our deployment, we decide to go by extending `easybuild/framework/extension.py` by adding properties such as : 
- `update_deps`
- `complete_deps`
- and provide a way to get `required_deps`

https://github.com/easybuilders/easybuild-framework/blob/e0e04fb773a5cee26669211f563696da94455cfd/easybuild/framework/extension.py#L238

Commit:
Refactor of udpate-exts-list

Commit:
Implementation of --update_exts_list for python, R and R-Bioconductor.
Implementation of --complete_exts_list for R and R-Bioconductor (missing Python).

Commit:
Implemetation of --complete-exts-list for R and R-Bioconductor. Pending for Python

Commit:
--complete-exts-list for python finished

Commit:
Forcing local_biocver parameter at easyconfig as biocondcutor's version setter

Commit:
Sanitize tool for R working fine

Commit:
small bugfix
